### PR TITLE
TextRank-fix

### PR DESCRIPTION
**Pull request by Gfreely**
fix #224 problem, update new UI to support change keyword extraction method.

**Main changes**

Added options to the RAG interface for selecting the keyword extraction method, along with several settings for TextRank, such as the number of keywords to extract, the size of the sliding window, and the processing language ('en' for English, 'zh' for Chinese).
QQ20250627-144754

A 'TextRank mask words' setting has also been added. It allows users to manually input specific phrases composed of letters and symbols to prevent them from being split during Chinese word segmentation. This requires manual entry by the user.
QQ20250627-144913

**Test results**

English
QQ20250627-151805
英文提取效果2

Chinese(Since I did not build the knowledge graph using Chinese, the queries used are different; however, the keyword extraction remains accurate)
中文提取效果
中文2

### CODE REVIEW

**Review by imbajin — COMMENTED** [review]
## Code Review Summary

**Overall Assessment**: Changes Required

### Key Findings
- 🚨 **Critical**: 3 issues found (breaking changes, error handling, security concerns)
- ⚠️ **Medium**: 4 improvements suggested  
- 🧹 **Minor**: 2 optimizations noted

### Main Concerns

#### 1. 🚨 Breaking API Changes Without Migration Path
The PR removes parameters (language, max_keywords) from method signatures in graph_rag_task.py without providing backward compatibility. This will break all existing code using these parameters.

**File**: hugegraph-llm/src/hugegraph_llm/operators/graph_rag_task.py (lines 57-82)
**Recommendation**: Add deprecated parameter support with warnings to maintain backward compatibility during transition period.

#### 2. 🚨 Insufficient Error Handling in TextRank Implementation
The new MultiLingualTextRank class lacks proper error handling for edge cases.

**File**: hugegraph-llm/src/hugegraph_llm/operators/document_op/textrank_word_extract.py (line 133)
**Issue**: Division by zero possible when max(pagerank_scores) is 0
**Fix**: Add zero check before normalization

#### 3. 🚨 Regex Injection Vulnerability
**File**: hugegraph-llm/src/hugegraph_llm/operators/document_op/textrank_word_extract.py (line 55)
The dynamic regex compilation from self.rules without proper sanitization could lead to ReDoS attacks.
**Fix**: Validate and sanitize regex patterns before compilation.

### Medium Priority Issues

#### 1. ⚠️ Uncaught Network Errors in NLTK Downloads
**File**: hugegraph-llm/src/hugegraph_llm/operators/common_op/nltk_helper.py (lines 53-57)
Only catching specific errors but not handling timeout or other network issues.

#### 2. ⚠️ Missing Input Validation
**File**: hugegraph-llm/src/hugegraph_llm/operators/llm_op/keyword_extract.py (lines 62-66)
The integer conversion for max_keywords catches exceptions but doesn't log warnings for invalid input.

#### 3. ⚠️ Hardcoded Magic Numbers
**File**: hugegraph-llm/src/hugegraph_llm/operators/document_op/textrank_word_extract.py (line 31)
Window size validation uses hardcoded max value (10) without explanation.

#### 4. ⚠️ Missing Tests
No test files included in this PR for the new TextRank functionality and hybrid extraction method.

### Minor Issues

#### 1. 🧹 Import Organization
Multiple files have import reorganization that seems unnecessary and makes the diff harder to read.

#### 2. 🧹 Inconsistent Language Handling
The language determination logic appears in multiple places - should be centralized.

### Positive Aspects
✅ Good addition of multiple extraction methods (LLM, TextRank, Hybrid)
✅ Proper logging added for debugging
✅ Config documentation updated appropriately

### Required Actions Before Approval
1. Add migration guide for breaking changes
2. Fix the division by zero issue in TextRank
3. Add input validation for regex patterns
4. Include comprehensive tests for new functionality
5. Consider adding backward compatibility layer

**Comment by imbajin** [comment]
## Detailed Code Review - Additional Comments

### Critical Issue Details

#### 1. Division by Zero in TextRank PageRank Normalization
In `textrank_word_extract.py`, line 129-131:
```python
pagerank_scores = self.graph.pagerank(directed=False, damping=0.85, weights='weight')
if max(pagerank_scores) > 0:
    pagerank_scores = [scores/max(pagerank_scores) for scores in pagerank_scores]
```

**Problem**: The code already checks for `max(pagerank_scores) > 0`, which is good, but doesn't handle the else case properly.

**Suggested Fix**:
```python
pagerank_scores = self.graph.pagerank(directed=False, damping=0.85, weights='weight')
if pagerank_scores and max(pagerank_scores) > 0:
    max_score = max(pagerank_scores)
    pagerank_scores = [score/max_score for score in pagerank_scores]
else:
    pagerank_scores = [0.0] * len(pagerank_scores) if pagerank_scores else []
```

#### 2. Breaking Changes in graph_rag_task.py

The removal of `language` and `max_keywords` parameters needs careful handling:

**Current problematic change** (lines 57-82):
```python
# Before:
def extract_keywords(self, text=None, max_keywords=5, language="english", extract_template=None):

# After: 
def extract_keywords(self, text=None, extract_template=None):
```

**Suggested backward-compatible implementation**:
```python
def extract_keywords(
    self,
    text: Optional[str] = None,
    extract_template: Optional[str] = None,
    max_keywords: Optional[int] = None,  # Deprecated
    language: Optional[str] = None,  # Deprecated
    **kwargs
):
    """
    Extract keywords from text.
    
    Args:
        text: Text to extract keywords from
        extract_template: Template for extraction
        max_keywords: [DEPRECATED] Use context['max_keywords'] instead
        language: [DEPRECATED] Use llm_settings.language instead
    """
    import warnings
    
    if max_keywords is not None:
        warnings.warn(
            "max_keywords parameter is deprecated and will be removed in v2.0. "
            "Please pass it via context dictionary instead.",
            DeprecationWarning,
            stacklevel=2
        )
        # Store in context for backward compatibility
        self._context['max_keywords'] = max_keywords
    
    if language is not None:
        warnings.warn(
            "language parameter is deprecated and will be removed in v2.0. "
            "Please configure it via llm_settings instead.",
            DeprecationWarning,
            stacklevel=2
        )
        # Store for backward compatibility
        self._context['language'] = language
        
    self._operators.append(
        KeywordExtract(
            text=text,
            extract_template=extract_template,
            **kwargs
        )
    )
    return self
```

#### 3. Regex Injection Prevention

In `textrank_word_extract.py`, the rules list should be immutable and validated:

**Current vulnerable code**:
```python
self.rules = [r"https?://\\S+|www\\.\\S+",
              r"\\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}\\b",
              r"\\b\\w+(?:[-'\\']\\w+)+\\b",
              r"\\b\\d+[,.]\\d+\\b"]
```

**Suggested secure implementation**:
```python
import re

class MultiLingualTextRank:
    # Make rules immutable
    _SAFE_RULES = tuple([
        r"https?://\\S+|www\\.\\S+",
        r"\\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}\\b",
        r"\\b\\w+(?:[-'\\']\\w+)+\\b",
        r"\\b\\d+[,.]\\d+\\b"
    ])
    
    def __init__(self, keyword_num: int = 5, window_size: int = 3):
        # Validate and compile patterns at initialization
        try:
            self._compiled_rules = re.compile('|'.join(self._SAFE_RULES), re.IGNORECASE)
        except re.error as e:
            raise ValueError(f"Invalid regex pattern: {e}")
```

#### 4. NLTK Error Handling Enhancement

In `nltk_helper.py`, add more comprehensive error handling:

```python
def check_nltk_data(self):
    """Check and download required NLTK data packages."""
    import socket
    
    # Set timeout for network operations
    original_timeout = socket.getdefaulttimeout()
    socket.setdefaulttimeout(30)
    
    try:
        # ... existing code ...
        for package in required_packages:
            try:
                # ... existing check ...
            except LookupError:
                max_retries = 3
                for attempt in range(max_retries):
                    try:
                        log.info("Downloading NLTK package %s (attempt %d/%d)", 
                                package, attempt + 1, max_retries)
                        nltk.download(package, download_dir=nltk_data_dir, quiet=False)
                        break
                    except (URLError, HTTPError, PermissionError, socket.timeout) as e:
                        if attempt == max_retries - 1:
                            log.error("Failed to download %s after %d attempts: %s", 
                                    package, max_retries, e)
                        else:
                            time.sleep(2 ** attempt)  # Exponential backoff
    finally:
        socket.setdefaulttimeout(original_timeout)
```

### Testing Requirements

The PR is missing test coverage for:

1. **TextRank algorithm tests**:
   - Test with empty text
   - Test with single word
   - Test with non-ASCII characters
   - Test window size edge cases
   - Test PageRank score normalization

2. **Hybrid extraction tests**:
   - Test weight balancing
   - Test fallback when one method fails
   - Test score combination logic

3. **Migration tests**:
   - Test deprecated parameter warnings
   - Test backward compatibility

Example test structure:
```python
def test_textrank_empty_input():
    extractor = MultiLingualTextRank()
    result = extractor.extract_keywords("")
    assert result == {}

def test_textrank_division_by_zero():
    extractor = MultiLingualTextRank()
    # Create scenario where all PageRank scores are 0
    result = extractor.extract_keywords("a a a")  
    assert all(score >= 0 for score in result.values())

def test_deprecated_parameters():
    with warnings.catch_warnings(record=True) as w:
        pipeline = RAGPipeline()
        pipeline.extract_keywords("test", max_keywords=5)
        assert len(w) == 1
        assert "deprecated" in str(w[0].message).lower()
```

File: hugegraph-llm/src/hugegraph_llm/operators/document_op/textrank_word_extract.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+        def _create_placeholder(match_obj):
+            nonlocal placeholder_id_counter
+            original_word = match_obj.group(0)
+            _placeholder = f" __shieldword_{placeholder_id_counter}__ "
+            placeholder_map[_placeholder.strip()] = original_word
+            placeholder_id_counter += 1
+            return _placeholder
+
+        special_regex = regex.compile('|'.join(self.rules), regex.V1)
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/operators/document_op/textrank_word_extract.py]: ⚠️ **Performance Issue: Repeated regex compilation**

The regex pattern is recompiled on every call to `_word_mask`, which is inefficient for repeated extractions.

**Recommendation:**
Compile once during initialization:
```python
def __init__(self, keyword_num: int = 5, window_size: int = 3):
    # ... existing code ...
    self.rules = [r"https?://\S+|www\.\S+", ...]
    self.special_regex = regex.compile('|'.join(self.rules), regex.V1)

def _word_mask(self, text):
    # ... existing code ...
    text = self.special_regex.sub(_create_placeholder, text)
```

File: hugegraph-llm/src/hugegraph_llm/config/llm_config.py [hunk 2]
```diff
@@ -30,6 +30,10 @@ class LLMConfig(BaseConfig):
     text2gql_llm_type: Literal["openai", "litellm", "ollama/local"] = "openai"
     embedding_type: Optional[Literal["openai", "litellm", "ollama/local"]] = "openai"
     reranker_type: Optional[Literal["cohere", "siliconflow"]] = None
+    keyword_extract_type: Literal["llm", "textrank", "hybrid"] = "llm"
+    window_size: Optional[int] = 3
+    hybrid_llm_weights: Optional[float] = 0.5
```
Comment (imbajin) [inline hugegraph-llm/src/hugegraph_llm/config/llm_config.py]: ⚠️ **Important: Missing validation for hybrid_llm_weights**

The config accepts `hybrid_llm_weights` but doesn't validate the 0.0-1.0 range at initialization.

**Recommendation:**
Add Pydantic field validation:
```python
from pydantic import Field

hybrid_llm_weights: Optional[float] = Field(
    default=0.5,
    ge=0.0,
    le=1.0,
    description="LLM weight in hybrid mode (0.0-1.0)"
)
```

**Review by imbajin — APPROVED** [review]
